File: src/poller.rs
use std::collections::{HashMap, HashSet};
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;
use tokio::sync::mpsc;
use tokio::task::JoinHandle;
use tokio::time::{sleep, Duration};

use crate::browser::BrowserExecutor;
use crate::callback::send_callback;
use crate::config::load_config;
use crate::models::{AppConfig, AppState, TaskData};

/// 掃單結果事件，由 poller 發送給 UI
#[derive(Debug)]
#[allow(dead_code)]
pub enum PollEvent {
    Log(String),
    TaskStarted(String),           // order_no
    TaskDone(String, bool),        // order_no, success
    Countdown(u64),                // 距下次掃單剩餘秒數（0=正在掃單）
    ActiveCount(usize),            // 當前活躍任務數
}

pub struct Poller {
    pub stop_flag: Arc<AtomicBool>,
    handle: Option<JoinHandle<()>>,
}

impl Poller {
    /// 啟動掃單循環，返回 Poller 句柄
    pub fn start(
        event_tx: mpsc::UnboundedSender<PollEvent>,
        state: Arc<std::sync::Mutex<AppState>>,
    ) -> Self {
        let stop_flag = Arc::new(AtomicBool::new(false));
        let stop_flag_clone = Arc::clone(&stop_flag);

        let handle = tokio::spawn(poll_loop(event_tx, state, stop_flag_clone));

        Poller {
            stop_flag,
            handle: Some(handle),
        }
    }

    /// 停止掃單
    pub fn stop(&self) {
        self.stop_flag.store(true, Ordering::Relaxed);
        if let Some(ref h) = self.handle {
            h.abort();
        }
    }
}

async fn poll_loop(
    event_tx: mpsc::UnboundedSender<PollEvent>,
    _state: Arc<std::sync::Mutex<AppState>>,
    stop_flag: Arc<AtomicBool>,
) {
    let log = |msg: &str| {
        let _ = event_tx.send(PollEvent::Log(msg.to_string()));
    };

    log("🔄 掃單服務已啟動");

    // 活躍任務：order_no -> JoinHandle
    let mut active_tasks: HashMap<String, JoinHandle<()>> = HashMap::new();
    // 正在運行的 order_no 集合
    let active_set: Arc<std::sync::Mutex<HashSet<String>>> =
        Arc::new(std::sync::Mutex::new(HashSet::new()));

    let http_client = reqwest::Client::new();

    loop {
        if stop_flag.load(Ordering::Relaxed) {
            break;
        }

        // 清理已完成的任務
        active_tasks.retain(|_, h| !h.is_finished());
        {
            let active_set_guard = active_set.lock().unwrap();
            let _ = event_tx.send(PollEvent::ActiveCount(active_set_guard.len()));
        }

        // 執行一次掃單
        let cfg = load_config();
        poll_once(
            &http_client,
            &cfg,
            &event_tx,
            &mut active_tasks,
            &active_set,
            &stop_flag,
        )
        .await;

        let interval = cfg.poll_interval;
        // 倒計時
        for remaining in (1..=interval).rev() {
            if stop_flag.load(Ordering::Relaxed) {
                break;
            }
            let _ = event_tx.send(PollEvent::Countdown(remaining));
            sleep(Duration::from_secs(1)).await;
        }
        if !stop_flag.load(Ordering::Relaxed) {
            let _ = event_tx.send(PollEvent::Countdown(0));
        }
    }

    // 停止所有活躍任務
    for (_, h) in active_tasks.drain() {
        h.abort();
    }

    log("⏹ 掃單服務已停止");
}

async fn poll_once(
    client: &reqwest::Client,
    cfg: &AppConfig,
    event_tx: &mpsc::UnboundedSender<PollEvent>,
    active_tasks: &mut HashMap<String, JoinHandle<()>>,
    active_set: &Arc<std::sync::Mutex<HashSet<String>>>,
    stop_flag: &Arc<AtomicBool>,
) {
    let log = |msg: String| {
        let _ = event_tx.send(PollEvent::Log(msg));
    };

    let server_url = cfg.server_url.trim_end_matches('/');
    if server_url.is_empty() || cfg.api_key.is_empty() {
        return;
    }

    let poll_url = format!("{server_url}/api/pcauto/pending-tasks");
    log(format!(">> 掃單 {poll_url}"));

    let t0 = std::time::Instant::now();
    let result = client
        .get(&poll_url)
        .header("X-Pcauto-Key", &cfg.api_key)
        .timeout(Duration::from_secs(10))
        .send()
        .await;

    let elapsed = t0.elapsed().as_secs_f64();

    match result {
        Err(e) if e.is_connect() || e.is_timeout() => {
            log(format!("❌ 無法連線後端（{elapsed:.2}s）：{poll_url}"));
        }
        Err(e) => {
            log(format!("❌ 掃單失敗（{elapsed:.2}s）: {e}"));
        }
        Ok(resp) => {
            let status = resp.status();
            if !status.is_success() {
                log(format!(
                    "⚠️ HTTP {status}（{elapsed:.2}s）—— 請確認 API Key 和後端 URL"
                ));
                return;
            }

            match resp.json::<crate::models::PendingTasksResponse>().await {
                Err(e) => {
                    log(format!("❌ 解析響應失敗: {e}"));
                }
                Ok(data) => {
                    let tasks = data.tasks;
                    if tasks.is_empty() {
                        log(format!("🔍 無待處理任務（{elapsed:.2}s）"));
                    } else {
                        log(format!("📋 發現 {} 個待處理任務（{elapsed:.2}s）", tasks.len()));
                    }

                    for task in tasks {
                        start_task(
                            client,
                            cfg,
                            task,
                            event_tx,
                            active_tasks,
                            active_set,
                            stop_flag,
                        )
                        .await;
                    }
                }
            }
        }
    }
}

async fn start_task(
    client: &reqwest::Client,
    cfg: &AppConfig,
    task: TaskData,
    event_tx: &mpsc::UnboundedSender<PollEvent>,
    active_tasks: &mut HashMap<String, JoinHandle<()>>,
    active_set: &Arc<std::sync::Mutex<HashSet<String>>>,
    global_stop: &Arc<AtomicBool>,
) {
    let order_no = task.order_no.clone();
    let task_id = task.task_id;

    // 避免重複啟動同一訂單
    {
        let guard = active_set.lock().unwrap();
        if guard.contains(&order_no) {
            return;
        }
    }
    // 如果舊 handle 還在跑也跳過
    if let Some(h) = active_tasks.get(&order_no) {
        if !h.is_finished() {
            return;
        }
    }

    let max_concurrent = cfg.max_concurrent_tasks.max(1);

    // 標記為活躍；如果已達並發上限，本輪暫不啟動新任務
    {
        let mut guard = active_set.lock().unwrap();
        if guard.len() >= max_concurrent {
            let _ = event_tx.send(PollEvent::Log(format!(
                "⏸ 已達並發上限 {max_concurrent}，暫不啟動任務: {order_no}"
            )));
            return;
        }
        guard.insert(order_no.clone());
    }

    let cfg_clone = cfg.clone();
    let event_tx_clone = event_tx.clone();
    let active_set_clone = Arc::clone(active_set);
    let global_stop_clone = Arc::clone(global_stop);
    let client_clone = client.clone();

    let _ = event_tx.send(PollEvent::Log(format!("🚀 已啟動任務: {order_no}")));
    let _ = event_tx.send(PollEvent::TaskStarted(order_no.clone()));

    let order_no_key = order_no.clone();

    let handle = tokio::spawn(async move {
        let log_tx_inner = {
            let (tx, mut rx) = mpsc::unbounded_channel::<String>();
            let evt_tx = event_tx_clone.clone();
            tokio::spawn(async move {
                while let Some(msg) = rx.recv().await {
                    let _ = evt_tx.send(PollEvent::Log(msg));
                }
            });
            tx
        };

        let task_stop = Arc::new(AtomicBool::new(false));
        // 同步 global stop 到 task stop
        let task_stop_clone = Arc::clone(&task_stop);
        let gs = Arc::clone(&global_stop_clone);
        tokio::spawn(async move {
            loop {
                if gs.load(Ordering::Relaxed) {
                    task_stop_clone.store(true, Ordering::Relaxed);
                    break;
                }
                tokio::time::sleep(Duration::from_millis(200)).await;
            }
        });

        let executor = BrowserExecutor::new(cfg_clone.clone(), task, log_tx_inner, task_stop);
        let success = executor.run().await;

        // 發送最終回調
        let status = if success { 2 } else { 3 };
        let cb_result = send_callback(&client_clone, &cfg_clone, task_id, &order_no, status).await;
        if let Err(e) = cb_result {
            let _ = event_tx_clone.send(PollEvent::Log(format!("❌ 回調失敗: {e}")));
        }

        // 從活躍集合移除
        {
            let mut guard = active_set_clone.lock().unwrap();
            guard.remove(&order_no);
        }

        let _ = event_tx_clone.send(PollEvent::TaskDone(order_no, success));
        let _ = event_tx_clone.send(PollEvent::ActiveCount(
            active_set_clone.lock().unwrap().len(),
        ));
    });

    active_tasks.insert(order_no_key, handle);
}

File: src/main/poller.ts
// Mirror of pcauto/src/poller.rs
//
// Topology (PLAN.md §6):
//   - setInterval-ish loop driven by `setTimeout` that calls /pending-tasks every
//     `cfg.poll_interval` seconds.
//   - For each new task, queue it onto `p-limit(cfg.max_concurrent_tasks)`.
//   - Stop = AbortController.signal.aborted; every long sleep / poll checks it.

import { BrowserWindow } from 'electron';
import pLimit from 'p-limit';
import { loadConfig } from './config';
import { sendCallback } from './api/callback';
import { httpRequest, safeJsonParse } from './api/http';
import { log } from './logger';
import { runTask } from './browser/executor';
import { parseTaskData, type PendingTasksResponse, type TaskData } from './types';
import { Ch } from '../shared/ipc-channels';

const CALLBACK_RETRY_DELAY_SECS = 10;
const CALLBACK_MAX_ATTEMPTS = 3;

function broadcastState<T>(channel: string, payload: T): void {
  for (const w of BrowserWindow.getAllWindows()) {
    if (!w.isDestroyed()) w.webContents.send(channel, payload);
  }
}

class Poller {
  private abort: AbortController | undefined;
  private loopPromise: Promise<void> | undefined;
  private activeOrders = new Set<string>();
  private limit: ReturnType<typeof pLimit> | undefined;

  isRunning(): boolean {
    return this.abort !== undefined;
  }

  start(): void {
    if (this.abort) return;
    this.abort = new AbortController();
    const cfg = loadConfig();
    if (!cfg.server_url || !cfg.api_key) {
      log('⚠️ 後端 URL 或 API Key 未設定,請先在設定中配置');
      this.abort = undefined;
      return;
    }
    this.limit = pLimit(Math.max(cfg.max_concurrent_tasks, 1));
    broadcastState(Ch.RunningState, { running: true });
    this.loopPromise = this.runLoop(this.abort.signal);
  }

  async stop(): Promise<void> {
    const ab = this.abort;
    if (!ab) return;
    ab.abort();
    this.abort = undefined;
    broadcastState(Ch.RunningState, { running: false });
    broadcastState(Ch.Countdown, { remaining: 0 });
    if (this.loopPromise) {
      try {
        await this.loopPromise;
      } catch {
        /* ignore */
      }
      this.loopPromise = undefined;
    }
  }

  private shouldStop(signal: AbortSignal): () => boolean {
    return () => signal.aborted;
  }

  private async runLoop(signal: AbortSignal): Promise<void> {
    log('🔄 掃單服務已啟動');
    try {
      while (!signal.aborted) {
        broadcastState(Ch.ActiveCount, { count: this.activeOrders.size });
        const cfg = loadConfig();
        await this.pollOnce(cfg, signal);
        const interval = Math.max(cfg.poll_interval, 1);
        for (let remaining = interval; remaining >= 1; remaining--) {
          if (signal.aborted) break;
          broadcastState(Ch.Countdown, { remaining });
          await new Promise<void>((r) => setTimeout(r, 1000));
        }
        if (!signal.aborted) broadcastState(Ch.Countdown, { remaining: 0 });
      }
    } finally {
      log('⏹ 掃單服務已停止');
    }
  }

  private async pollOnce(
    cfg: ReturnType<typeof loadConfig>,
    signal: AbortSignal,
  ): Promise<void> {
    const serverUrl = cfg.server_url.replace(/\/+$/, '');
    if (!serverUrl || !cfg.api_key) return;
    const pollUrl = `${serverUrl}/api/pcauto/pending-tasks`;
    log(`>> 掃單 ${pollUrl}`);
    const t0 = Date.now();
    try {
      const resp = await httpRequest(pollUrl, {
        method: 'GET',
        headers: { 'X-Pcauto-Key': cfg.api_key },
        timeoutMs: 10_000,
        signal,
      });
      const elapsed = ((Date.now() - t0) / 1000).toFixed(2);
      if (!resp.ok) {
        log(`⚠️ HTTP ${resp.status}(${elapsed}s) —— 請確認 API Key 和後端 URL`);
        return;
      }
      const data = safeJsonParse<PendingTasksResponse>(resp.text);
      if (!data) {
        log('❌ 解析響應失敗');
        return;
      }
      const rawTasks = Array.isArray(data.tasks) ? data.tasks : [];
      if (rawTasks.length === 0) {
        log(`🔍 無待處理任務(${elapsed}s)`);
      } else {
        log(`📋 發現 ${rawTasks.length} 個待處理任務(${elapsed}s)`);
      }
      for (const raw of rawTasks) {
        let task: TaskData;
        try {
          task = parseTaskData(raw);
        } catch (e) {
          log(`❌ 解析任務失敗: ${(e as Error).message}`);
          continue;
        }
        this.startTask(task, cfg.max_concurrent_tasks, signal);
      }
    } catch (e) {
      const elapsed = ((Date.now() - t0) / 1000).toFixed(2);
      log(`❌ 掃單失敗(${elapsed}s): ${(e as Error).message}`);
    }
  }

  private startTask(task: TaskData, maxConcurrent: number, signal: AbortSignal): void {
    const orderNo = task.order_no;
    if (this.activeOrders.has(orderNo)) return;
    const cap = Math.max(maxConcurrent, 1);
    if (this.activeOrders.size >= cap) {
      log(`⏸ 已達並發上限 ${cap},暫不啟動任務: ${orderNo}`);
      return;
    }
    this.activeOrders.add(orderNo);
    log(`🚀 已啟動任務: ${orderNo}`);
    broadcastState(Ch.TaskStarted, { order_no: orderNo });
    broadcastState(Ch.ActiveCount, { count: this.activeOrders.size });

    const limit = this.limit;
    if (!limit) {
      // shouldn't happen, but fall back to inline execution
      void this.runTaskWithCallback(task, signal).finally(() => {
        this.activeOrders.delete(orderNo);
        broadcastState(Ch.ActiveCount, { count: this.activeOrders.size });
      });
      return;
    }

    void limit(() => this.runTaskWithCallback(task, signal))
      .catch(() => {
        /* swallow — runTaskWithCallback already logs */
      })
      .finally(() => {
        this.activeOrders.delete(orderNo);
        broadcastState(Ch.ActiveCount, { count: this.activeOrders.size });
      });
  }

  private async runTaskWithCallback(task: TaskData, signal: AbortSignal): Promise<void> {
    const cfg = loadConfig();
    const shouldStop = this.shouldStop(signal);
    let success = false;
    try {
      const out = await runTask({ cfg, task, shouldStop });
      if (out.status === 'stopped') {
        broadcastState(Ch.TaskDone, { order_no: task.order_no, success: false });
        return;
      }
      const isSuccess = out.status === 'success';
      const callback = isSuccess
        ? {
            status: 2,
            reason: 'pcauto 自動化全部步驟已完成,準備完成訂單',
            label: '完成',
            failure_image_data: undefined as string | undefined,
          }
        : {
            status: 3,
            reason: out.failure?.reason ?? '未知失敗',
            label: '失敗',
            failure_image_data: out.failure?.failure_image_data,
          };
      success = await this.sendFinalCallback(task, callback, shouldStop);
      if (isSuccess && !success) {
        // backend rejected the success callback after all retries
      }
    } catch (e) {
      log(`❌ 任務內部錯誤: ${(e as Error).message}`);
    } finally {
      broadcastState(Ch.TaskDone, { order_no: task.order_no, success });
    }
  }

  private async sendFinalCallback(
    task: TaskData,
    cb: {
      status: number;
      reason: string;
      label: string;
      failure_image_data: string | undefined;
    },
    shouldStop: () => boolean,
  ): Promise<boolean> {
    const cfg = loadConfig();
    let attempt = 0;
    while (!shouldStop()) {
      attempt += 1;
      try {
        const ok = await sendCallback({
          cfg,
          taskId: task.task_id,
          orderNo: task.order_no,
          status: cb.status,
          reason: cb.reason,
          failureImageData: cb.failure_image_data,
        });
        if (ok) return cb.status === 2;
        if (attempt >= CALLBACK_MAX_ATTEMPTS) {
          log(
            `❌ ${cb.label}回調被後端拒絕;已失敗 ${attempt}/${CALLBACK_MAX_ATTEMPTS} 次,釋放任務`,
          );
          return false;
        }
        log(
          `❌ ${cb.label}回調被後端拒絕;第 ${attempt}/${CALLBACK_MAX_ATTEMPTS} 次失敗,${CALLBACK_RETRY_DELAY_SECS}s 後重試,暫不釋放任務`,
        );
      } catch (e) {
        if (attempt >= CALLBACK_MAX_ATTEMPTS) {
          log(
            `❌ ${cb.label}回調失敗: ${(e as Error).message};已失敗 ${attempt}/${CALLBACK_MAX_ATTEMPTS} 次,釋放任務`,
          );
          return false;
        }
        log(
          `❌ ${cb.label}回調失敗: ${(e as Error).message};第 ${attempt}/${CALLBACK_MAX_ATTEMPTS} 次失敗,${CALLBACK_RETRY_DELAY_SECS}s 後重試`,
        );
      }
      for (let i = 0; i < CALLBACK_RETRY_DELAY_SECS; i++) {
        if (shouldStop()) return false;
        await new Promise<void>((r) => setTimeout(r, 1000));
      }
    }
    return false;
  }
}

export const poller = new Poller();
